x/tally/keeper/endblock.go

```diff
@@ -209,44 +209,57 @@ func (k Keeper) FilterAndTally(ctx sdk.Context, req types.Request) (TallyResult,
 		sort.Strings(reveals[i].ProxyPubKeys)
 	}
 
-	result.execGasUsed = calculateExecGasUsed(reveals)
-
-	filter, err := base64.StdEncoding.DecodeString(req.ConsensusFilter)
-	if err != nil {
-		return result, k.logErrAndRet(ctx, err, types.ErrDecodingConsensusFilter, req)
-	}
-	// Convert base64-encoded payback address to hex encoding that
-	// the tally VM expects.
-	decodedBytes, err := base64.StdEncoding.DecodeString(req.PaybackAddress)
+	// Phase I: Filtering
+	filter, err := k.BuildFilter(ctx, req.ConsensusFilter, req.ReplicationFactor)
 	if err != nil {
-		return result, k.logErrAndRet(ctx, err, types.ErrDecodingPaybackAddress, req)
+		return result, err
 	}
-	paybackAddrHex := hex.EncodeToString(decodedBytes)
-
-	filterResult, err := k.ApplyFilter(ctx, filter, reveals, req.ReplicationFactor)
+	filterResult, err := ApplyFilter(filter, reveals)
 	result.consensus = filterResult.Consensus
 	result.proxyPubKeys = filterResult.ProxyPubKeys
 	if err != nil {
 		return result, k.logErrAndRet(ctx, err, types.ErrApplyingFilter, req)
 	}
```

Comment:
This is not correct, as we should still calculate the exec gas used and determine the payouts in the case of consensus on a non-zero exit code.

Reply:
Ah right, thanks. I refactored the way we handle errors in this function. For that matter though, we ALWAYS want to calculate exec gas & payouts right? That is, we go through phase III even when there is an error while building filter or when there is no consensus.

Reply:
> That is, we go through phase III even when there is an error while building filter or when there is no consensus.

Yes, not sure how/if we do exec gas calculations though, but there should be some sort of payouts I guess. Gets even more fun if we still want to pay out data proxies :P